File: pkg-r/R/dep-discovery.R
# Automatic renderer HTML-dependency discovery (#146, #203).
#
# R's `page_react_html()` is a plain UI value: it is rendered before any
# session's `server()` runs, so — unlike Python's Express-mode
# `set_react_page()` — output dependencies cannot be inlined into the initial
# page. Instead, after every reactive flush we diff the session's registered
# outputs, extract each new output's UI via `output_ui()`, and push any
# not-yet-sent dependencies to the client as a `shinyreact-deps` custom
# message. The JS bundle loads them and re-runs `Shiny.bindAll()`; the shiny
# client replays stored output values on late bind, and hidden-output
# suspension means most values are only computed after the bind anyway.
#
# Diffing on *every* flush (not just the first) also covers outputs registered
# after startup — e.g. a module server mounted inside an observer.
#
# The hook: the JS bundle sends one `.shinyreact_init` ping (type
# `shinyreact.init`) after Shiny initializes; that type's dedicated input
# handler (input-handler.R) calls `install_dep_discovery()`. Every session
# gets exactly one ping, whether or not the app has any other inputs.

install_dep_discovery <- function(session) {
  if (is.null(session)) {
    return(invisible(FALSE))
  }
  user_data <- tryCatch(session$userData, error = function(e) NULL)
  if (!is.environment(user_data)) {
    return(invisible(FALSE))
  }
  if (isTRUE(user_data[[".shinyreact_dep_discovery"]])) {
    return(invisible(FALSE))
  }
  # `.outputs` is private; shiny exposes no API to enumerate registered
  # outputs (Python reads `session.output._outputs`, same deal). getOutput()
  # is public. MockShinySession has neither -- discovery no-ops there.
  private <- tryCatch(session$.__enclos_env__$private, error = function(e) NULL)
  if (
    is.null(private) ||
      !is.function(session$onFlushed) ||
      !is.function(session$getOutput)
  ) {
    return(invisible(FALSE))
  }
  user_data[[".shinyreact_dep_discovery"]] <- TRUE

  seen_outputs <- character()
  sent_deps <- character()
  push_new_output_deps <- function() {
    output_names <- names(private$.outputs)
    new_names <- setdiff(output_names, seen_outputs)
    if (length(new_names) == 0L) {
      return(invisible())
    }
    seen_outputs <<- c(seen_outputs, new_names)

    deps <- list()
    for (name in new_names) {
      ui <- output_ui_or_null(session$getOutput(name), name)
      if (!is.null(ui)) {
        deps <- c(deps, htmltools::findDependencies(ui))
      }
    }
    deps <- htmltools::resolveDependencies(deps)
    keys <- vapply(
      deps,
      function(dep) paste0(dep$name, "@", dep$version),
      character(1)
    )
    deps <- deps[!(keys %in% sent_deps)]
    if (length(deps) == 0L) {
      return(invisible())
    }
    sent_deps <<- c(sent_deps, setdiff(keys, sent_deps))

    # createWebDependency() registers each dep's resource path with shiny so
    # the client can fetch the files; the client skips deps already on the
    # page by name, so overlap with the static <head> is harmless.
    session$sendCustomMessage(
      "shinyreact-deps",
      lapply(deps, shiny::createWebDependency)
    )
  }
  session$onFlushed(push_new_output_deps, once = FALSE)
  invisible(TRUE)
}

File: pkg-py/src/shinyreact/_dep_discovery.py
"""Automatic renderer HTML-dependency discovery for Core mode (#146, #203, #220).

Express's ``set_react_page()`` can inline renderer dependencies into the page
head because the page function runs after the renderers mount. A Core-mode page
(:func:`page_react`, :func:`page_react_html`) is built before ``server()`` runs,
so there is nothing to inline. Instead — the same design R uses
(``pkg-r/R/dep-discovery.R``) — after every reactive flush we diff the session's
registered outputs, extract each new output's UI, and push any not-yet-sent
dependencies to the client as a ``shinyreact-deps`` custom message. The JS
bundle loads them and re-runs ``Shiny.bindAll()``.

Diffing on *every* flush (not just the first) also covers outputs registered
after startup — e.g. a module server mounted inside an observer.

The hook: the JS bundle sends one ``.shinyreact_init`` ping (type
``shinyreact.init``) after Shiny initializes; that type's input handler
(``_input_handler.py``) calls :func:`install_dep_discovery`. Every session gets
exactly one ping, whether or not the app has any other inputs.
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any, cast

from htmltools import HTMLDependency, Tag, TagList

if TYPE_CHECKING:
    from shiny.session import Session

_INSTALLED_FLAG = "_shinyreact_dep_discovery"


def install_dep_discovery(session: Session | None) -> bool:
    """Install the per-session flush hook. Returns whether it was installed."""
    if session is None:
        return False
    # `_outputs` and `on_flushed` are missing on mock/express-stub sessions;
    # discovery no-ops there rather than raising.
    outputs = getattr(getattr(session, "output", None), "_outputs", None)
    if outputs is None or not callable(getattr(session, "on_flushed", None)):
        return False
    # Two copies of the JS bundle on one page send two pings; install once.
    # (py-shiny has no `session$userData` equivalent, so: an attribute.)
    if getattr(session, _INSTALLED_FLAG, False):
        return False
    setattr(session, _INSTALLED_FLAG, True)

    seen_outputs: set[str] = set()
    sent_deps: set[str] = set()

    async def push_new_output_deps() -> None:
        new_names = [name for name in outputs if name not in seen_outputs]
        if not new_names:
            return
        seen_outputs.update(new_names)

        deps: list[HTMLDependency] = []
        for name in new_names:
            ui = outputs[name].renderer.auto_output_ui()
            if isinstance(ui, (Tag, TagList)):
                deps.extend(ui.tagify().get_dependencies())
        deps = [d for d in deps if f"{d.name}@{d.version}" not in sent_deps]
        if not deps:
            return
        sent_deps.update(f"{d.name}@{d.version}" for d in deps)

        # `_process_ui()` registers each dep's resource route with the app and
        # returns the client-side JSON (Python's `createWebDependency()`). The
        # client skips deps already on the page, so overlap is harmless.
        payload = session._process_ui(TagList(*deps))["deps"]
        await session.send_custom_message(
            "shinyreact-deps", cast("dict[str, object]", payload)
        )

    session.on_flushed(cast(Any, push_new_output_deps), once=False)
    return True
